### PR TITLE
docs: Corellium sandbox

### PR DESCRIPTION
**Pull request by pawelpasterz**
This PR introduces naive corellium client implementation + sandbox with an example script.
There is lots of kotlin code, yet it's not the main purpose of this task, just keep it in mind during the review 😅 

#### All necessary info is included in documentation MD files:
* [corellium-client](https://github.com/Flank/flank/blob/1607-corellium-playground/corellium/corellium-client/README.md)
* [corellium-sandbox](https://github.com/Flank/flank/blob/1607-corellium-playground/corellium/corellium-sandbox/README.md)
* [SUMMARY](https://github.com/Flank/flank/blob/1607-corellium-playground/corellium/corellium-sandbox/SUMMARY.md)

#### Actions:
- [x] Login
- [x] Get Project Ids
- [x] Get Project info
- [x] Get Instances
- [x] Get Instance info
- [x] Create Instance
- [x] Upload file
- [x] Get VPN config
- [x] properties file
- [x] Retry mechanism
- [x] Configurable logging
- [x] Documentation -- client
- [x] Documentation -- sandbox
- [x] Documentation -- summary
- [x] Run script

### CODE REVIEW

**Comment by pawelpasterz** [comment]
@Mergifyio refresh

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

File: corellium/corellium-client/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+repositories {
+    jcenter()
+    mavenCentral()
+    maven(url = "https://kotlin.bintray.com/kotlinx")
+}
+
+tasks.withType<KotlinCompile> { kotlinOptions.jvmTarget = "1.8" }
+
+dependencies {
+    implementation(kotlin("stdlib", org.jetbrains.kotlin.config.KotlinCompilerVersion.VERSION)) // or "stdlib-jdk8"
```
Comment (piotradamczyk5) [inline corellium/corellium-client/build.gradle.kts:17]: strange dependency

File: corellium/corellium-client/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    implementation(Dependencies.KOTLIN_SERIALIZATION)
+    implementation("io.ktor:ktor-client-serialization:1.5.1")
+    implementation("io.ktor:ktor-client-websockets:1.5.1")
+    implementation("io.ktor:ktor-client-logging:1.5.1")
+
+    implementation("ch.qos.logback:logback-classic:1.2.3")
+
+    api("io.ktor:ktor-client-core:1.5.1")
+    api("io.ktor:ktor-client-cio:1.5.1")
```
Comment (piotradamczyk5) [inline corellium/corellium-client/build.gradle.kts:26]: should not we add them to `Dependencies` file ?

File: corellium/corellium-client/build.gradle.kts (hunk 1 above)
Comment (pawelpasterz) [reply corellium/corellium-client/build.gradle.kts:17]: Yet we have it in a couple of places among repo

File: corellium/corellium-client/build.gradle.kts (hunk 2 above)
Comment (pawelpasterz) [reply corellium/corellium-client/build.gradle.kts:26]: Sure we can 👍

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Comment by piotradamczyk5** [comment]
@pawelpasterz generally speaking this PR is hard to review and check. Configuration requires many manual steps to review it.
Could we somehow create a docker image for testing Corellium with VPN and other stuff already setup?

**Comment by pawelpasterz** [comment]
> @pawelpasterz generally speaking this PR is hard to review and check. Configuration requires many manual steps to review it.
> Could we somehow create a docker image for testing Corellium with VPN and other stuff already setup?
> 
> As a volunteer I would like to help with this

Well, that's definitely beyond the scope of this task. As I mentioned several times, this is a more research task rather than a fully working implementation.

Within this PR I've created a very basic background for future work and continued investigation.

In the `SUMMARY` you can see a couple of `action needed` headers. Those are my comments and proposals for improvements in the flow. It may turn out that a docker image could help in one of the. Yet, as I mentioned earlier --  it's beyond the task scope. If it weren't, it should have the estimation like 100 or more, rather than 3-5...

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Comment by piotradamczyk5** [comment]
> > @pawelpasterz generally speaking this PR is hard to review and check. Configuration requires many manual steps to review it.
> > Could we somehow create a docker image for testing Corellium with VPN and other stuff already setup?
> > As a volunteer I would like to help with this
> 
> Well, that's definitely beyond the scope of this task. As I mentioned several times, this is a more research task rather than a fully working implementation. In other words -- what we need to learn to be able to integrate, not -- how to integrate
> 
> Within this PR I've created a very basic background for future work and continued investigation.
> 
> In the `SUMMARY` you can see a couple of `action needed` headers. Those are my comments and proposals for improvements in the flow. It may turn out that a docker image could help in one of the. Yet, as I mentioned earlier -- it's beyond the task scope. If it weren't, it should have the estimation like 100 or more, rather than 3-5...
> 
> Regarding docker, I am not sure if that would help that much, you need to set up a VPN only once. It would be great to have it automagically set up and be downloadable for everyone -- that's true. Yet, again, it was not the purpose of this task.

So my basic question is should we review docs or test implementation or both of them?

**Comment by pawelpasterz** [comment]
> So my basic question is should we review docs or test implementation or both of them?
I am a little bit confused if this task is about providing documentation or POC or both of them that's why I asked about docker image for testing.
We could always create a follow-up task for it

You are right, https://github.com/Flank/flank/issues/1607 is a bit outdated and can be misleading. This is more a POC, minimum steps that allow us to run iOS tests on corellium instance without the node-based client.

As a result of it, we are going to create a bunch of other issues/epics. (see action needed sections in the summary)

I assume not a single line from my client implementation will be used in the final version. So you can review the code, but it most likely won't be used anywhere. It's up to you.

We can commit directly to the master branch (and that's a good thing!) that's why I do it with PR

I hope that answers your question :)

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Review by piotradamczyk5 — APPROVED** [review]

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Comment by Sloox** [comment]
@pawelpasterz is this reviewable or are we shelving this until further notice?

**Review by Sloox — CHANGES_REQUESTED** [review]
@pawelpasterz is this reviewable or are we shelving this until further notice?

**Comment by pawelpasterz** [comment]
> @pawelpasterz is this reviewable or are we shelving this until further notice?

Yes, please review! I want finally close this PR :)

I suppose you might have similar questions as Pitor had, please check https://github.com/Flank/flank/pull/1621#issuecomment-791348624 (and older comments if you like)

File: corellium/corellium-client/README.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
+## About
+
+Only for POC integration purposes. It's based on [corellium-api](https://github.com/corellium/corellium-api). It has
+only essential features implemented to be able to run simple iOS test run similar
```
Comment (Sloox) [inline corellium/corellium-client/README.md:4]: The wording here is unclear?
Can you check it again.

Comment (Sloox) [reply corellium/corellium-client/README.md:4]: Possibly:
```suggestion
only essential features implemented to be able to run a simple set of IOS tests that coincide 
```

File: corellium/corellium-client/README.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+```
+
+#### createNewInstance(newInstance: Instance): String (instance id)
+
+Creates new [Instance](#data-class-instance), returns newly created instance id.
+
+NOTE: instance created with this method is not ready to use (state ==  `creating`).
+Use [waitUntilInstanceIsReady](#waituntilinstanceisreadyinstanceid-string) to ensure VM has `on` state
+
+Following options are required (all `Instance` fields are described [here](#data-class-instance))
```
Comment (Sloox) [inline corellium/corellium-client/README.md:89]:
```suggestion
The following options are required (all `Instance` fields are described [here](#data-class-instance))
```

File: corellium/corellium-client/README.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+Returns a list of [Instance](#data-class-instance) created for the project with the given id
+
+```kotlin
+val projectId = "1111-2222-3333-aaa-bbbbbb-cccc"
+val instances: List<Instance> = client.getProjectInstancesList(projectId)
+```
+
+#### createNewInstance(newInstance: Instance): String (instance id)
+
+Creates new [Instance](#data-class-instance), returns newly created instance id.
```
Comment (Sloox) [inline corellium/corellium-client/README.md:84]:
```suggestion
Creates a new [Instance](#data-class-instance), returns newly created instance id.
```

**Review by Sloox — APPROVED** [review]
Very minor comments

**Comment by pawelpasterz** [comment]
@Mergifyio rebase

**Pull request merged**